### PR TITLE
[Websocket] Update options parameter to headers. Update to spec.

### PR DESCRIPTION
**Pull request by christopherdro**
This is a follow up of https://github.com/facebook/react-native/commit/9b87e6c860a95fe3d55285314d3d56be06cb7833.

### PR Overview:
- Allows custom headers on connection request
- Adds a default `origin` header to Android, just like iOS

**Introduces no breaking changes.**

### PR Detail:

I was working on something similar and would like to propose a few changes that make the API more consistent across both iOS and Android platforms and brings this closer to [spec](https://tools.ietf.org/html/rfc6455).

I believe @aprock first implementation of adding custom `headers` was correct. It makes sense naming this argument `headers` since we have no other general options available, and the current `options` field is being used to pass in a header anyway.

My use case for custom headers was attaching a token to the `Authorization` header on the connection request. I have been testing this by passing a JWT inside the `Authorization` header and verifying it on the server before establishing a connection.

I tried to take this this abstraction further and read the response headers from the server to see if a new token was available if the connection was denied from a expired token. Unfortunately, this was only possible with Android since the third party library used by iOS (SquareRocket) does not return the server response on connection failures.

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 1]
```diff
@@ -187,4 +200,31 @@ private void notifyWebSocketFailed(int id, String message) {
...
+      if (requestURI.getPort() != -1) {
+        defaultOrigin = String.format("%s://%s:%s", scheme, requestURI.getHost(), requestURI.getPort());
+      } else {
+        defaultOrigin = String.format("%s://%s/", scheme, requestURI.getHost());
+      }
+
+      return defaultOrigin;
+
+    } catch(URISyntaxException e) {
+      // Handle error
```
Comment (christopherdro) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: @mkonicek @satya164 Not sure the best way to handle the error here. Ideas?

Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: @christopherdro Since this is a user error, we should throw the error to show a redbox (`IllegalArgumentException`). This matches the behaviour on the web where it throws an error.

**Comment by christopherdro** [comment]
@satya164 I updated the PR to handle the error message. However, it appears that passing a invalid url scheme will be caught beforehand through [Request.Builder](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java#L76)

**Comment by christopherdro** [comment]
@satya164 Should probably update this if this gets merged in.
https://github.com/facebook/react-native/commit/8f19f5bef4a86c046fd9dd0ad5501f6bb5bbd6c4

File: Libraries/WebSocket/RCTSRWebSocket.h [hunk 2]
```diff
@@ -61,11 +61,11 @@ extern NSString *const RCTSRHTTPResponseErrorKey;
 
 // Protocols should be an array of strings that turn into Sec-WebSocket-Protocol.
 // options can contain a custom "origin" NSString
-- (instancetype)initWithURLRequest:(NSURLRequest *)request protocols:(NSArray<NSString *> *)protocols options:(NSDictionary<NSString *, NSString *> *)options NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithURLRequest:(NSURLRequest *)request protocols:(NSArray<NSString *> *)protocols headers:(NSDictionary<NSString *, NSString *> *)headers NS_DESIGNATED_INITIALIZER;
```
Comment (mkonicek) [inline Libraries/WebSocket/RCTSRWebSocket.h]: Ah this used to be called options but it's really HTTP headers?

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 3]
```diff
@@ -492,7 +492,22 @@ - (void)didConnect
...
+    [_requestedHeaders enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
+       if ([obj isKindOfClass:[NSString class]]) {
+         CFHTTPMessageSetHeaderFieldValue(request, (__bridge CFStringRef)key, (__bridge CFStringRef)obj);
+       } else {
+         RCTSRFastLog(@"Ignoring passed in header: %@, value not a string.", key);
+       }
+     }];
+  } else {
+    CFHTTPMessageSetHeaderFieldValue(request, CFSTR("origin"), (__bridge CFStringRef)(_url.RCTSR_origin));
+  }
```
Comment (mkonicek) [inline Libraries/WebSocket/RCTSRWebSocket.m]: cc @nicklockwood As my Obj-C fu is not great :)

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 4]
```diff
@@ -75,14 +77,25 @@ public void connect(final String url, @Nullable final ReadableArray protocols, @
...
-        builder.addHeader("Origin", options.getString("origin"));
-      } else {
-        FLog.w(
-          ReactConstants.TAG,
-          "Ignoring: requested origin, value not a string");
+    if (headers != null) {
+      ReadableMapKeySetIterator iterator = headers.keySetIterator();
+
+      if (!headers.hasKey("origin")) {
+        builder.addHeader("origin", setDefaultOrigin(url));
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: n00b question: should this be "Origin" via the spec?

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 5]
```diff
@@ -75,14 +77,25 @@ public void connect(final String url, @Nullable final ReadableArray protocols, @
...
       }
+
+      while (iterator.hasNextKey()) {
+        String key = iterator.nextKey();
+        if (ReadableType.String.equals(headers.getType(key))) {
+          builder.addHeader(key, headers.getString(key));
+        } else {
+          FLog.w(
+            ReactConstants.TAG,
+            "Ignoring: requested origin, value not a string");
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: `iterator` loops over all the headers, correct? Why log "requested origin" here even if that specific item is unrelated to the origin header?

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 6]
```diff
@@ -187,4 +200,30 @@ private void notifyWebSocketFailed(int id, String message) {
     params.putString("message", message);
     sendEvent("websocketFailed", params);
   }
+
+  private static String setDefaultOrigin(String uri) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: I cannot tell what this does based on the name. Can you add short Javadoc please?

File: Libraries/WebSocket/RCTSRWebSocket.h (hunk 2 above)
Comment (christopherdro) [inline Libraries/WebSocket/RCTSRWebSocket.h]: Yea, these was recently merged via https://github.com/facebook/react-native/commit/9b87e6c860a95fe3d55285314d3d56be06cb7833

I noticed that his original implementation did have an option for using headers but was asked to remove it because only the `origin` header was required by the contributor.

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java (hunk 5 above)
Comment (christopherdro) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: My bad, origin should be replaced there by the key name. Will update shortly.

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java (hunk 4 above)
Comment (christopherdro) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: Http headers are case-insensitive and get lowercased automatically when being down with the request.

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 7]
```diff
@@ -245,7 +245,7 @@ + (void)initialize;
   CRLFCRLF = [[NSData alloc] initWithBytes:"\r\n\r\n" length:4];
 }
 
-- (instancetype)initWithURLRequest:(NSURLRequest *)request protocols:(NSArray<NSString *> *)protocols options:(NSDictionary<NSString *, NSString *> *)options
+- (instancetype)initWithURLRequest:(NSURLRequest *)request protocols:(NSArray<NSString *> *)protocols headers:(NSDictionary<NSString *, NSString *> *)headers
```
Comment (nicklockwood) [inline Libraries/WebSocket/RCTSRWebSocket.m]: Since NSURLRequest already includes headers, is there any reason for the seperate headers parameter? Can't we just include them in the request?

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 8]
```diff
@@ -492,7 +492,22 @@ - (void)didConnect
     CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Protocol"), (__bridge CFStringRef)[_requestedProtocols componentsJoinedByString:@", "]);
   }
 
-  CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Origin"), (__bridge CFStringRef)(_requestedOptions[@"origin"] ?: _url.RCTSR_origin));
+  if (_requestedHeaders) {
+    
+    if (![_requestedHeaders objectForKey:@"origin"]) {
```
Comment (nicklockwood) [inline Libraries/WebSocket/RCTSRWebSocket.m]: if (!_requestedHeaders[@"origin"]) {

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 9]
```diff
@@ -492,7 +492,22 @@ - (void)didConnect
...
+      CFHTTPMessageSetHeaderFieldValue(request, CFSTR("origin"), (__bridge CFStringRef)(_url.RCTSR_origin));
+    }
+    
+    [_requestedHeaders enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
+       if ([obj isKindOfClass:[NSString class]]) {
```
Comment (nicklockwood) [inline Libraries/WebSocket/RCTSRWebSocket.m]: Could just use `[obj description]` to convert it to a string here. That will do the right thing for numbers, which are the most likey kind of wrong value to get passed in.

File: Libraries/WebSocket/RCTWebSocketModule.m [hunk 10]
```diff
@@ -44,9 +44,9 @@ - (void)dealloc
   }
 }
 
-RCT_EXPORT_METHOD(connect:(NSURL *)URL protocols:(NSArray *)protocols options:(NSDictionary *)options socketID:(nonnull NSNumber *)socketID)
+RCT_EXPORT_METHOD(connect:(NSURL *)URL protocols:(NSArray *)protocols headers:(NSDictionary *)headers socketID:(nonnull NSNumber *)socketID)
 {
-  RCTSRWebSocket *webSocket = [[RCTSRWebSocket alloc] initWithURL:URL protocols:protocols options:options];
+  RCTSRWebSocket *webSocket = [[RCTSRWebSocket alloc] initWithURL:URL protocols:protocols headers:headers];
```
Comment (nicklockwood) [inline Libraries/WebSocket/RCTWebSocketModule.m]: We can just merge the url and headers into an NSURLRequest at this point, and use the other constructor.

    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL];
    request.allHTTPHeaderFields = headers;

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 11]
```diff
@@ -187,4 +200,37 @@ private void notifyWebSocketFailed(int id, String message) {
     params.putString("message", message);
     sendEvent("websocketFailed", params);
   }
+
+  /**
+   * Set a default origin
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: Can you please explain what setting a default origin means?

**Comment by mkonicek** [comment]
We talked about this briefly at React.js conf but don't remember the conclusion. Regarding the `setDefaultOriginMethod` - why is the origin header needed in React Native? I thought it was designed for browsers?

**Comment by christopherdro** [comment]
@mkonicek Your are right that it is browser specfic. I was just trying to make things consistent across platform. If it's unnecessary then we should remove it from iOS as well.

@nicklockwood Thoughts on whether we need to have a default origin on iOS?

**Comment by mkonicek** [comment]
Cool, thanks for the explanation! I vote for getting rid of the origin headers if they're not needed on mobile.

File: Libraries/WebSocket/RCTSRWebSocket.m (hunk 7 above)
Comment (christopherdro) [inline Libraries/WebSocket/RCTSRWebSocket.m]: @nicklockwood That makes sense but it appears there are two other places that headers are being set.
1. https://github.com/facebook/react-native/blob/debcac59d06d527c3a6a07ef9561fe4041022ff4/Libraries/WebSocket/RCTSRWebSocket.m#L478-L489
2. https://github.com/facebook/react-native/blob/debcac59d06d527c3a6a07ef9561fe4041022ff4/Libraries/WebSocket/RCTSRWebSocket.m#L298

So im not sure the best way to handle this.

**Comment by christopherdro** [comment]
@nicklockwood Updated the PR based on some of your comments and made further notes on ones that I wasn't sure about.

**Comment by mkonicek** [comment]
Is the method `setDefaultOrigin` needed? Do we need to set an origin header?

**Comment by christopherdro** [comment]
Technically no since RN is not a browser client.

> The |Origin| header field [RFC6454] is used to protect against
   unauthorized cross-origin use of a WebSocket server by scripts using
   the WebSocket API in a web browser.  The server is informed of the
   script origin generating the WebSocket connection request.  If the
   server does not wish to accept connections from this origin, it can
   choose to reject the connection by sending an appropriate HTTP error
   code.  This header field is sent by browser clients; for non-browser
   clients, this header field may be sent if it makes sense in the
   context of those clients.

I was waiting to see what @nicklockwood thought about removing the default origin header from iOS.

**Comment by christopherdro** [comment]
@nicklockwood Any chance you can chime in here again so I can go ahead and wrap up this PR? Thanks!

**Comment by christopherdro** [comment]
@mkonicek Fair enough. I'll update the the PR to only update the headers. Should have it done by this weekend and will ping you again.

**Comment by zxcpoiu** [comment]
Hi @christopherdro,

What is your opinion to keep the argument name `options`, and use 'headers' as a key for setting up additional header ?

**ex: **
```
options = {
    headers: {}
    otherOptions: ...
}
```

**Comment by christopherdro** [comment]
@zxcpoiu I believe that was my original thought but avoided since it seems that the general attitude of PR's is to implement whats required atm and not over engineer. I will however revisit this shortly.

Can you provide any examples of what some future options might be?

**Comment by zxcpoiu** [comment]
Hi @christopherdro 

you're right, **not over engineer** is a good point.

Currently what I can think of potential use cases are:  maybe to tweak options of android [OkHttpClient Consturrctor](http://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html) when [create ws instance](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java#L68)

**for example:**
* read/write Timeout
* self-signed certificate verify
* force TLS versions

It's not that critical, and I don't want to hold this decent PR.
Just treat it as a note here for reference :)

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit
